File: Python/DFS/BurningTree.py
# @author
# Aakash Verma

# Output:
# 2
# 7 4 5
# 6 3
# 1 
# 0 8 

from collections import deque

class Node:
    def __init__(self, data):
        self.data = data
        self.right = self.left = None

class BurningTree:
    
    def __init__(self): 
        self.root = None

    queue = deque([])

    def burningTree(self, root, targetNode):
        if root is None:
            return 0
        if root.data == targetNode:
            print(root.data)
            if root.left is not None:
                self.queue.append(root.left)
            if root.right is not None:
                self.queue.append(root.right)
            return 1

        left = self.burningTree(root.left, targetNode)
        if(left == 1):
            queueSize = len(self.queue)
            for i in range(queueSize):
                current = self.queue.popleft()
                print(current.data, end = " ")
                if current.left is not None:
                    self.queue.append(self.current.left)
                if current.right is not None:
                    self.queue.append(self.current.right)

            if root.right is not None:
                self.queue.append(root.right)
            print(root.data)
            return 1

        right = self.burningTree(root.right, targetNode)
        if(right == 1):
            queueSize = len(self.queue)
            for i in range(queueSize):
                current = self.queue.popleft()
                print(current.data, end = " ")
                if current.left is not None:
                    self.queue.append(self.current.left)
                if current.right is not None:
                    self.queue.append(self.current.right)

            if root.left is not None:
                self.queue.append(root.left)
            print(root.data)
            return 1

        
if __name__ == '__main__':
    tree = BurningTree()
    tree.root = Node(3)
    tree.root.left = Node(5)
    tree.root.right = Node(1)
    tree.root.left.left = Node(6)
    tree.root.left.right = Node(2)
    tree.root.right.left = Node(0)
    tree.root.right.right = Node(8)
    tree.root.left.right.left = Node(7)
    tree.root.left.right.right = Node(4)

    tree.burningTree(tree.root, 2)
    while tree.queue:
        queueSize = len(tree.queue)
        for i in range(queueSize):
            current = tree.queue.popleft()
            print(current.data, end = " ")
            if current.left is not None:
                tree.queue.append(current.left)
            if current.right is not None:
                tree.queue.append(current.right)
        print()

    

    

File: Java/DFS/BurningTree.java
/*
*   @author
*   Aakash Verma
*
*   Output:
*   2
*   7 4 5
*   6 3
*   1 
*   0 8 
*
*/

import java.util.*;

class Node {
    int data;
    Node left, right;
    public Node(int key) {
        data = key;
        left = right = null;
    }
}


class BurningTree {

    static Node root;
    BurningTree() {
        root = null;
    }

    private static Queue<Node> queue = new LinkedList<>();

    private static int burningTree(Node root, int targetNode) {
        if(root == null)
            return 0;
        if(root.data == targetNode) {
            System.out.println(root.data);
            if(root.left != null)
                queue.offer(root.left);
            if(root.right != null)
                queue.offer(root.right);
            return 1;
        }

        int left = burningTree(root.left, targetNode);
        if(left == 1) {
            int queueSize = queue.size();
            for(int i = 0; i < queueSize; i++) {
                Node current = queue.poll();
                System.out.print(current.data + " ");
                if(current.left != null)
                    queue.offer(current.left);
                if(current.right != null)
                    queue.offer(current.right);
            }
            if(root.right != null) {
                queue.offer(root.right);
            }
            System.out.println(root.data);
            return 1;
        }

        int right = burningTree(root.right, targetNode);
        if(right == 1) {
            int queueSize = queue.size();
            for(int i = 0; i < queueSize; i++) {
                Node current = queue.poll();
                System.out.print(current.data + " ");
                if(current.left != null)
                    queue.offer(current.left);
                if(current.right != null)
                    queue.offer(current.right);
            }
            if(root.left != null) {
                queue.offer(root.left);
            }
            System.out.println(root.data);
            return 1;
        }
        return 0;
    }

    public static void main (String[] args) {
        root = new Node(3);
        root.left = new Node(5);
        root.right = new Node(1);
        root.left.left = new Node(6);
        root.left.right = new Node(2);
        root.right.left = new Node(0);
        root.right.right = new Node(8);
        root.left.right.left = new Node(7);
        root.left.right.right = new Node(4);
        burningTree(root, 2);
        while(!queue.isEmpty()) {
            int queueSize = queue.size();
            for(int i = 0; i < queueSize; i++) {
                Node current = queue.poll();
                System.out.print(current.data + " ");
                if(current.left != null)
                    queue.offer(current.left);
                if(current.right != null)
                    queue.offer(current.right);
            }
            System.out.println();
        }
    }
}